### DISCUSSION TITLE
Automating store_selection_keys

### DISCUSSION THREAD

**Discussion by tinny-blueleo** [L3MON4D3/LuaSnip#584] [General]
Hi friends!  I have mappings for generating snippets via lua functions, I'm trying to automate the process of wrapping lines in snippets like if statements or for loops if I have the line highlighted with line V mode.  My store_selection_keys is set to `<C-s>` and in my function I try to automate the storing of the selection before expanding the snippet by trying to send `<C-s>` to the buffer like so:

`vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes("<C-s>", true, false, true), vim.fn.mode(), true)`

However when the snippet expands it doesn't expand the snippet around the line that I wanted.  Anyone know if there is any API instead to store the selection or a better way to send the keys?  Manually pressing Ctrl+s works before expanding the snippet.

**Comment by L3MON4D3** [comment]
Ouh, feedkeys is a bit difficult, because it will not immediately execute, I'd recommend reading `:h feedkeys` (the mode you're supposed to pass is not the vim-mode, but corresponds to certain feedkeys-behaviour, it's explained in the help). It's possible to work around that by queueing the lua-commands with the same feedkey-call, the downside here is that it gets pretty unreadable quickly.

We don't have an api for storing visual, we did not want to manually implement deleting the selection (here's the [implementation](https://github.com/L3MON4D3/LuaSnip/blob/3a814e069c73d49887b708417a9f58b948e807da/lua/luasnip/config.lua#L203)).
(If you always delete entire lines, you could also manually call `util.store_selection()` and just delete the line yourself (`nvim_buf_set_text`))

**Reply by tinnyw** [reply]
This will work perfectly for my case, thanks for pointing me in the right direction L3MON4D3!

**Comment by tinnyw** [comment]
Hi L3MON4D3,

Sorry to bug you about this again.  I'm trying to utilize store_selection but random lines are being stored in the selection and not necessarily what is highlighted before being in the snippet:

`if vim.api.nvim_get_mode().mode == "V" or vim.api.nvim_get_mode().mode == "v" then
    require('luasnip.util.util').store_selection()
    vim.cmd(":normal dd")
end

ls.snip_expand(snippet)`

I was trying to copy over the function directly from the implementation but it was throwing a syntax error around the <.

`vim.cmd([[:lua require('luasnip.util.util').store_selection()<cr>gv"_s]])` and this wasn't working either for the store selection call.

**Reply by leiserfg** [reply]
Well, the `<cr>gv"_s` is part of a mapping, it does not work standalone in a cmd.

**Reply by leiserfg** [reply]
Maybe using `nvim_feedkeys` will do it.
